persistence/sql/persister_recovery.go

```diff
@@ -127,3 +128,67 @@ func (p *Persister) DeleteExpiredRecoveryFlows(ctx context.Context, expiresAt ti
 	}
 	return nil
 }
+
+func (p *Persister) CreateRecoveryCode(ctx context.Context, recoveryCode *code.RecoveryCode) error {
+	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.CreateRecoveryCode")
+	defer span.End()
+
+	code := recoveryCode.Code
+	recoveryCode.Code = p.hmacValue(ctx, code)
+	recoveryCode.NID = p.NetworkID(ctx)
+
+	// This should not create the request eagerly because otherwise we might accidentally create an address that isn't
+	// supposed to be in the database.
+	if err := p.GetConnection(ctx).Create(recoveryCode); err != nil {
+		return err
+	}
+
+	recoveryCode.Code = code
+	return nil
+}
+
+func (p *Persister) UseRecoveryCode(ctx context.Context, fID uuid.UUID, codeVal string) (*code.RecoveryCode, error) {
+	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.UseRecoveryCode")
+	defer span.End()
+
+	var recoveryCode code.RecoveryCode
+
+	nid := p.NetworkID(ctx)
+	if err := sqlcon.HandleError(p.Transaction(ctx, func(ctx context.Context, tx *pop.Connection) (err error) {
+		for _, secret := range p.r.Config().SecretsSession(ctx) {
+			if err = tx.Where("code = ? AND nid = ? AND NOT used AND selfservice_recovery_flow_id = ?", p.hmacValueWithSecret(ctx, codeVal, secret), nid, fID).First(&recoveryCode); err != nil {
+				if !errors.Is(sqlcon.HandleError(err), sqlcon.ErrNoRows) {
+					return err
+				}
+			} else {
+				break
+			}
+		}
+		if err != nil {
+			return err
+		}
+
+		var ra identity.RecoveryAddress
+		if err := tx.Where("id = ? AND nid = ?", recoveryCode.RecoveryAddressID, nid).First(&ra); err != nil {
```

Comment:
Could this instead be done with a JOIN above? You could also use a `RETURNING` clause to combine all three queries into one.

Reply:
I guess it might be possible, but not sure if it would actually be an improvement especially with the current system of iterating through the secrets?

Reply:
`IN` queries seem to be very difficult with `pop`. I saw `sqlx.In` but that is targeted at full queries, IDK how it works with `pop`. Why is iterating over the secrets an issue? 

>You could also use a RETURNING clause to combine all three queries into one.

`RETURNING` is not supported on `mysql`...

Reply:
The solution is OK!